src/info/git/mod.rs

```diff
@@ -0,0 +1,321 @@
+    let mut time_of_most_recent_commit = None;
+    let mut time_of_first_commit = None;
+    let mut number_of_commits_by_signature: HashMap<Sig, usize> = HashMap::new();
+    let mailmap = repo.open_mailmap();
+    let bot_regex_pattern = get_no_bots_regex(&options.info.no_bots)?;
+    let has_commit_graph_traversal_ended = Arc::new(AtomicBool::default());
+    let total_number_of_commits = Arc::new(AtomicUsize::default());
+    let commit_graph = repo.commit_graph().ok();
+    let commit_iter = repo
+        .head_commit()?
+        .id()
+        .ancestors()
+        .sorting(Sorting::ByCommitTimeNewestFirst)
+        .use_commit_graph(commit_graph.is_some())
+        .with_commit_graph(commit_graph)
+        .all()?;
+    let (author_threads, author_tx) = get_author_channel(repo, &bot_regex_pattern, &mailmap);
+    let (churn_thread, churn_tx) = get_churn_channel(
+        repo,
+        &has_commit_graph_traversal_ended,
+        &total_number_of_commits,
+        options.info.churn_pool_size,
+    )?;
+
+    let mut count = 0;
+    for commit in commit_iter {
+        let commit = commit?;
+        {
+            if options.info.no_merges && commit.parent_ids.len() > 1 {
+                continue;
+            }
+
+            author_tx.send(commit.id)?;
+            let commit_time = gix::actor::Time::new(
+                commit
+                    .commit_time
+                    .expect("sorting by time yields this field as part of traversal")
+                    as u32, // TODO: remove this cast once `gix` supports 64 bit dates.
+                0,
+            );
+            time_of_most_recent_commit.get_or_insert(commit_time);
+            time_of_first_commit = commit_time.into();
+
+            count += 1;
+        }
+
+        churn_tx.send(commit.id).ok();
+    }
+
+    drop(author_tx);
+    for thread in author_threads {
+        let mapping = thread.join().expect("no panic")?;
+        for (sig, num_commits) in mapping {
+            *number_of_commits_by_signature.entry(sig).or_insert(0) += num_commits;
+        }
+    }
+
+    total_number_of_commits.store(count, Ordering::SeqCst);
+    has_commit_graph_traversal_ended.store(true, Ordering::SeqCst);
+
+    let (number_of_commits_by_file_path, churn_pool_size) =
+        churn_thread.join().expect("never panics")?;
+
+    let git_metrics = GitMetrics::new(
+        number_of_commits_by_signature,
+        number_of_commits_by_file_path,
+        churn_pool_size,
+        count,
+        time_of_first_commit,
+        time_of_most_recent_commit,
+        repo.is_shallow(),
+        options,
+    )?;
+
+    Ok(git_metrics)
+}
+
+fn get_author_channel(
+    repo: &gix::Repository,
+    bot_regex_pattern: &Option<MyRegex>,
+    mailmap: &gix::mailmap::Snapshot,
+) -> (
+    Vec<JoinHandle<Result<HashMap<Sig, usize>>>>,
+    Sender<ObjectId>,
+) {
+    let num_threads = std::thread::available_parallelism()
```

Comment:
Isn't there a risk of thread starvation, if we use up all the available threads here. What about the thread used for compute churns?